### PR TITLE
FloatTexture version Blendshape 

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

This PR mainly include:
- Use `float texture mode` when `BlendShape` count is limit by GPU attribute count
- When there are no normals and tangents, the number of `attribute mode` support is increased to 8

Else:
- Add `Texture2DArraty` 
- Unified texture name, rename `TextureCubeMap` to `TextureCube`
- Add `ListenerUpdateFlag` and rename `UpdateFlag` to `BoolUpdateFlag`

### CODE REVIEW

File: packages/core/src/mesh/BlendShapeManager.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+  _releaseMemoryCache(): void {
+    this._layoutDirtyFlag.destroy();
+    const dataChangedFlags = this._subDataDirtyFlags;
+    for (let i = 0, n = dataChangedFlags.length; i < n; i++) {
+      dataChangedFlags[i].destroy();
+    }
+
+    this._layoutDirtyFlag = null;
+    this._subDataDirtyFlags = null;
+    this._blendShapes = null;
```
Comment (zhuxudong) [inline packages/core/src/mesh/BlendShapeManager.ts:323]: how can i get blendshape name if you release it

**Review by zhuxudong — CHANGES_REQUESTED** [review]

File: packages/core/src/mesh/BlendShapeManager.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/mesh/BlendShapeManager.ts:332]: done

File: packages/core/src/mesh/BlendShapeManager.ts [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+      return true;
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _needUpdateData(): boolean {
+    for (let i = 0, n = this._subDataDirtyFlags.length; i < n; i++) {
+      if (this._subDataDirtyFlags[i].flag) {
```
Comment (singlecoder) [inline packages/core/src/mesh/BlendShapeManager.ts:109]: this._subDataDirtyFlags is called repeatedly

File: packages/core/src/mesh/BlendShapeManager.ts [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    let textureHeight = 1;
+    if (textureWidth > maxTextureSize) {
+      textureHeight = Math.ceil(textureWidth / maxTextureSize);
+      textureWidth = maxTextureSize;
+    }
+
+    const shapeCount = this._blendShapes.length;
+    let blendShapeDataTexture = this._dataTexture;
+    const blendShapeCount = this._blendShapes.length;
+
```
Comment (singlecoder) [inline packages/core/src/mesh/BlendShapeManager.ts:245]: What is the difference between shapeCount and blendShapeCount

**Review by singlecoder — APPROVED** [review]
+1

File: packages/core/src/texture/Texture2DArray.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+   * @param imageSource - The source of texture
+   * @param flipY - Whether to flip the Y axis
+   * @param premultiplyAlpha - Whether to premultiply the transparent channel
+   * @param x - X coordinate of area start
+   * @param y - Y coordinate of area start
+   */
+  setImageSource(
+    elementIndex: number,
+    imageSource: TexImageSource | OffscreenCanvas,
+    mipLevel: number = 0,
```
Comment (yangfengzzz) [inline packages/core/src/texture/Texture2DArray.ts:98]: comment

File: packages/core/src/texture/Texture2DArray.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      x,
+      y
+    );
+  }
+
+  /**
+   * Get pixel color buffer.
+   * @param out - Color buffer
+   */
+  getPixelBuffer(elementIndex: number, out: ArrayBufferView): void;
```
Comment (yangfengzzz) [inline packages/core/src/texture/Texture2DArray.ts:119]: elementIndex comment

**Review by yangfengzzz — CHANGES_REQUESTED** [review]

File: packages/core/src/mesh/BlendShapeManager.ts [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        offset += 12;
+        elementCount += 3;
+      }
+      if (this._useBlendTangent) {
+        vertexElements.push(new VertexElement(`TANGENT_BS${i}`, offset, VertexElementFormat.Vector3, 0));
+        elementCount += 3;
+      }
+    }
+
+    this._lastUpdateLayoutAndCountInfo.setValue(this._blendShapeCount, +this._useBlendNormal, +this._useBlendTangent);
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/mesh/BlendShapeManager.ts:137]: convert bool to vector3

File: packages/core/src/mesh/BlendShapeManager.ts [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+    const blendShapeCount = this._blendShapes.length;
+
+    blendShapeDataTexture && blendShapeDataTexture.destroy();
+
+    blendShapeDataTexture = new Texture2DArray(
+      this._engine,
+      textureWidth,
+      textureHeight,
+      blendShapeCount,
+      TextureFormat.R32G32B32A32,
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/mesh/BlendShapeManager.ts:254]: A little hard to support float16 buffer:https://stackoverflow.com/questions/32633585/how-do-you-convert-to-half-floats-in-javascript

**Review by yangfengzzz — APPROVED** [review]

**Pull request merged**